### PR TITLE
Domain-Knowledge PhaseNet  (DKPN) picker  integration

### PR DESCRIPTION
**Pull request by mbagagli**
Inclusion of Domain-Knowledge PhaseNet (DKPN) picker as published in Lomax et al. 2024 (https://doi.org/10.26443/seismica.v3i1.1164) and some pre-trained weights.

The code has been refactored from its original repository (https://github.com/INGV/DKPN) to suit v0.8 of seisbench (originally developed over v0.3/0.4).

Everything seems fine, but probably double checks are needed by the mantainer(s).

The end-user is referred to the original paper for details and to the original repository for the training scripts.

Additional bug fixes for `base.WaveformModel` included that double check the right device for the machine at line 489-492 during execution of the `from_pre_trained` method. This should cover cases where model-weights are trained/stored over a GPU but later used on a CPU-only machine."

### CODE REVIEW

**Comment by yetinam** [comment]
Hi @mbagagli ,
thanks a lot for submitting this PR! I'll have a look at it in the new year. Happy holidays!

File: seisbench/models/base.py [hunk 1]
```diff
@@ -486,7 +486,10 @@ def load(cls, path, version_str=None):
         with open(path_json, "r") as f:
             weights_metadata = json.load(f)
         # Load model weights
-        model_weights = torch.load(f"{path_pt}")
+        #MB# : double check the right device for the machine
+        _dev = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (yetinam) [inline seisbench/models/base.py:490]: Thanks for the fix. I'd suggest a small change, in that I think by default we should load models always to CPU. Users can still move them to GPU manually afterwards. As our models are small, this doesn't have any performance downsides and it will keep the GPU usage explicit.

File: seisbench/models/dkpn.py [hunk 2]
```diff
@@ -0,0 +1,784 @@
...
+from seisbench.models.base import WaveformModel, _cache_migration_v0_v3
+
+# ---------  For PreProc class
+from obspy.core import Trace
+import copy
+from obspy.signal.filter import bandpass
+from scipy.signal import lfilter
+
+
+class DKPN(WaveformModel):
```
Comment (yetinam) [inline seisbench/models/dkpn.py:20]: Given this is essentially a PhaseNet model, it should inherit from PhaseNet

File: seisbench/models/dkpn.py [hunk 3]
```diff
@@ -0,0 +1,784 @@
...
+    """
+    .. document_args:: seisbench.models PhaseNet
```
Comment (yetinam) [inline seisbench/models/dkpn.py:22]: This should read `.. document_args:: seisbench.models DKPN`

File: seisbench/models/dkpn.py [hunk 4]
```diff
@@ -0,0 +1,784 @@
...
+    """
+
+    _annotate_args = WaveformModel._annotate_args.copy()
```
Comment (yetinam) [inline seisbench/models/dkpn.py:25]: `default_args` should be specified (and documented) here, so that they appear correctly in the documentation.

File: seisbench/models/dkpn.py [hunk 5]
```diff
@@ -0,0 +1,784 @@
...
+                2 * filters, filters, self.kernel_size, padding="same", bias=False
+            )
+            bn2 = nn.BatchNorm1d(filters, eps=1e-3)
+
+            self.up_branch.append(nn.ModuleList([conv_up, bn1, conv_same, bn2]))
+
+        self.out = nn.Conv1d(last_filters, self.classes, 1, padding="same")
+        self.softmax = torch.nn.Softmax(dim=1)
+
+        # DKPN attributes for taking care of windows CF and probs
```
Comment (yetinam) [inline seisbench/models/dkpn.py:157]: Any transient attributes should not be stored in class variables. From what I see, these variable are actually only written to but never read. Can they be deleted?

File: seisbench/models/dkpn.py [hunk 6]
```diff
@@ -0,0 +1,784 @@
...
+        self.batch_ = []
+        self.batch_cfs = []
+        self.batch_probs = []
+        self.stream_cfs = None
+
+    def __reset_predict(self):
```
Comment (yetinam) [inline seisbench/models/dkpn.py:163]: See comment above

File: seisbench/models/dkpn.py [hunk 7]
```diff
@@ -0,0 +1,784 @@
...
+    # ==================================  PROCESS STREAM
+    # ================================================================
+
+    def annotate_stream_pre(self, stream, argdict):
+        """ ARGDICT represent the default dict,merged (eventually)
+            with the key:value pairs of the "*.json" specifying the model
+        """
+        self.__reset_predict()
+
+        # -----------------   Pre-pend the original code of SEISBENCH (v0.2)
```
Comment (yetinam) [inline seisbench/models/dkpn.py:219]: I think you can just call `stream = super().annotate_stream_pre(stream, argdict)` instead.

File: seisbench/models/dkpn.py [hunk 8]
```diff
@@ -0,0 +1,784 @@
...
+            "blinding", self._annotate_args.get("blinding")[1]
+        )
+        if prenan > 0:
+            batch[:, :prenan] = np.nan
+        if postnan > 0:
+            batch[:, -postnan:] = np.nan
+        self.batch_probs.append(batch.detach().clone())
+        return batch
+
+    def extract_cf(self, stream):
```
Comment (yetinam) [inline seisbench/models/dkpn.py:275]: The function and the two below are never called.

File: seisbench/models/dkpn.py [hunk 9]
```diff
@@ -0,0 +1,784 @@
...
+            del model_args[key]
+
+        model_args["in_channels"] = self.in_channels
+        model_args["classes"] = self.classes
+        model_args["phases"] = self.labels
+        model_args["sampling_rate"] = self.sampling_rate
+
+        return model_args
+
+    def get_defaults(self):
```
Comment (yetinam) [inline seisbench/models/dkpn.py:340]: Given `default_args` is a public variable, I think it's unnecessary adding a wrapper function.

File: seisbench/models/dkpn.py [hunk 10]
```diff
@@ -0,0 +1,784 @@
...
+        return self.default_args
+
+    def set_dkpn_parameter(self, pardict):
```
Comment (yetinam) [inline seisbench/models/dkpn.py:343]: ... or a setter function for this mapper.

File: seisbench/models/dkpn.py [hunk 11]
```diff
@@ -0,0 +1,784 @@
...
+
+    _annotate_args = WaveformModel._annotate_args.copy()
+    _annotate_args["*_threshold"] = ("Detection threshold for the provided phase", 0.3)
+    _annotate_args["blinding"] = (
+        "Number of prediction samples to discard on each side of each window prediction",
+        (0, 0),
+    )
+    _annotate_args["overlap"] = (_annotate_args["overlap"][0], 1500)
+
+    _weight_warnings = [
```
Comment (yetinam) [inline seisbench/models/dkpn.py:33]: This is unnecessary here. It warns about outdated weights versions but for DKNP we do not have these.

File: seisbench/models/dkpn.py [hunk 12]
```diff
@@ -0,0 +1,784 @@
...
+# ====================================================================
+# ====================================================================
+# ====================================================================
+# ====================================================================
+
+
+class PreProc(object):
+    def __init__(self, **kwargs):
+        self.stream = None  # --> Instantiated everytime in the work() class-method
+        self.__dict__.update(kwargs)
```
Comment (yetinam) [inline seisbench/models/dkpn.py:436]: This works, but could you explicitly define the arguments in the function header and safe them explicitly. This will make documentation and debugging much easier.

File: seisbench/models/dkpn.py [hunk 13]
```diff
@@ -0,0 +1,784 @@
...
+# ====================================================================
+# ====================================================================
+# ====================================================================
+# ====================================================================
+# ====================================================================
+
+
+class PreProc(object):
+    def __init__(self, **kwargs):
+        self.stream = None  # --> Instantiated everytime in the work() class-method
```
Comment (yetinam) [inline seisbench/models/dkpn.py:435]: As for the DKPN class, temporary variables should not be saved in class variables. Instead, you could make `work` return the processed stream.

File: seisbench/models/dkpn.py [hunk 14]
```diff
@@ -0,0 +1,784 @@
...
+        Provides an augmentation for seisbench.generate.generator.GenericGenerator.add_augmentations().
+        """
+
+        labels, metadata_labels = state_dict["y"]
+        waveforms, metadata_waveforms = state_dict["X"]  # metadata is the relevant row-field of CSV
+        waveforms = self.__matrix_cfs__(waveforms)
+
+        # ================================  Emulates WINDOW-LEVEL NORM
+        # --- Remove FP stabilization
+        fstab = int(metadata_waveforms["trace_sampling_rate_hz"]*self.fp_stabilization)
```
Comment (yetinam) [inline seisbench/models/dkpn.py:450]: I'm not sure I understand these lines. I guess they should remove some padding in the beginning, but they also truncate at the end?

File: seisbench/models/dkpn.py [hunk 15]
```diff
@@ -0,0 +1,784 @@
...
+
+        # --- NORMALIZE  MODULUS matrix with the its STD
+        waveforms[4, :] = waveforms[4, :] / (
+            np.std(waveforms[4, :], axis=-1, keepdims=True) + 1e-10
+            )
+
+        state_dict["X"] = (waveforms, metadata_waveforms)  # OVERRIDE MATRIX --> OUTPUT
+        state_dict["y"] = (labels, metadata_labels)  # OVERRIDE MATRIX --> OUTPUT
+
+    def __matrix_cfs__(self, MM):
```
Comment (yetinam) [inline seisbench/models/dkpn.py:467]: I think this should be `_matrix_cfs` instead as private functions in python typically start with a single underscore and don't use trailing underscores.

File: seisbench/models/dkpn.py [hunk 16]
```diff
@@ -0,0 +1,784 @@
...
+        fp_band_data = []
+        fp_max_arrays_band = []
+        fp_max_arrays_val = []
+
+        for _idx, arr in enumerate(MM):
+
+            # ====================
+            # 1) Run FBsummary
+            summary = FBSummary(arr,
+                                sampling_rate=100.0,
```
Comment (yetinam) [inline seisbench/models/dkpn.py:489]: Is it intended that the sampling rate is hard-coded here?

File: seisbench/models/dkpn.py [hunk 17]
```diff
@@ -0,0 +1,784 @@
...
+
+        # --- Bring everything to float32
+        fp_cf_waveforms = fp_cf_waveforms.astype("float32")
+
+        return fp_cf_waveforms
+
+    def get_stream(self):
+        return self.stream
+
+    def work(self, stream):
```
Comment (yetinam) [inline seisbench/models/dkpn.py:547]: From what I can see, this function assumes that the stream consists of exactly three traces, named `ZNE`, from a single station, and aligned in time. This is not given for streams passed to `annotate`. Instead, this work should be done in `annotate_batch_pre`.

File: seisbench/models/dkpn.py [hunk 18]
```diff
@@ -0,0 +1,784 @@
...
+                self.stream += Trace(data=_cfdata,
+                                     header=_cfstats)
+
+        # # ==================== FP_ STABILIZATION
+        # print("DEBUG: removing the first %.2f seconds of trace" % self.fp_stabilization)
+        # for _tr in self.stream:
+        #     _tr.trim(_tr.stats.starttime + self.fp_stabilization,
+        #              _tr.stats.endtime)
+
+    def PolarizationFP(self, fp_band_data, fp_max_arrays_band_amax,
```
Comment (yetinam) [inline seisbench/models/dkpn.py:602]: Function names should be snake case and start with a non-capital letter.

File: seisbench/models/dkpn.py [hunk 19]
```diff
@@ -0,0 +1,784 @@
...
+        for j in range(n_bands):
+            octave_high = (self.freqmin + self.freqmin * 2.0) / 2.0 * (2**j)
+            octave_low = octave_high / 2.0
+            BF[j] = bandpass(self.data, octave_low, octave_high,
+                             self.sampling_rate,
+                             corners=self.cnr,
+                             zerophase=False)
+        return BF
+
+    def get_summary(self):
```
Comment (yetinam) [inline seisbench/models/dkpn.py:741]: I think the wrapper can be deleter.

**Comment by yetinam** [comment]
Hi @mbagagli ,
now I got to look through your PR. Thanks again for the contribution! There are a few points that need to be adjusted:

- The class copies a lot of code from PhaseNet. Following the DRY principle, instead the class should inherit from PhaseNet and just overwrite the necessary parts. This will also make reviewing easier.
- The `annotate` function does not work (at least for me). I've put a comment in the code where it crashes. The problem is that your modified inputs aren't correctly passed in. I think, you should do the following: (1) in `annotate_stream_pre` calculate the FilterPicker traces. From what I understood this is done per component, so it's pretty straightforward. Note that at this stage SeisBench has not aligned the different traces. Write the outputs into the stream with components like `FPA`, `FPB`, `FPC`. (2) Set the `component_order` to `ZNEABC`. This will force the model to cut give 6 channel data to `annotate_batch_pre`. (3) In `annotate_batch_pre`, calculate modulus and inclination. At this stage, SeisBench has aligned the traces, split them into different stations, ... . This means, the calculations are simple matrix operations. Afterwards, discard the channels you don't actually want to pass to the model (ZNE I guess). With these three steps, everything should work.
- Currently, the implementation saved temporary data into class variables, making the class stateful. Unless absolutely necessary, this should be avoided.
- To keep the size of the repository smaller, no weights are stored directly in the repository. However, I'll upload them to the SeisBench repository to make them generally available.
- Please add a unit test for DKPN. They can be very simple (see the examples already implemented), just checking that if you pass a valid input stream, the output of  `annotate` is not empty.
- Please run the [precommit hooks](https://github.com/seisbench/seisbench/blob/main/CONTRIBUTING.md#submitting-a-pull-request) to adjust the formatting.

Let me know if you have any questions and I'll be happy to help.
